### PR TITLE
[WIP] Postgres' hstore support

### PR DESCRIPTION
**Pull request by arteam**
This PR is to track progress and getting a feedback for implementing a support for Postgres' `hstore` type.

### CODE REVIEW

File: postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class HStoreArgumentFactory implements ArgumentFactory {
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Optional<Argument> build(Type type, Object value, StatementContext ctx) {
+        if (value instanceof Map) {
+            // Unfortunately, the PostgreSQL driver doesn't support mapping from Java maps to the hstore type.
+            // Therefore, we convert the map to a string representation and pass it to a PostgreSQL server.
+            return Optional.of((i, p, cx) -> p.setObject(i, toString((Map<String, ?>) value), Types.OTHER));
```
Comment (qualidafial) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: The Postgres driver supports `hstore` out of the box. You can just call `setObject` with the map:

    statement.setObject(col, map);

Also the Postgres driver thoroughly escapes keys and values--see `HStoreConverter` in the Postgres JDBC jar.

Alternatively you could use `PGobject`:

    PGobject obj = new PGobject();
    obj.setType("hstore");
    obj.setValue(HStoreConverter.toString(map));
    statement.setObject(col, obj);

File: postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        }
+        return Optional.empty();
+    }
+
+    private static String toString(Map<String, ?> map) {
+        return map.entrySet().stream()
+                .map(e -> "\"" + e.getKey() + "\" => \"" + e.getValue() + "\"")
```
Comment (stevenschlansker) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: Probably need some kind of escaping here, or at least an assertion that the key / value don't contain `"`

File: postgres/src/main/java/org/jdbi/v3/postgres/HStoreColumnMapper.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * A column mapper which maps Postgres' hstore type to Java's {@link Map}.
+ */
+public class HStoreColumnMapper implements ColumnMapper<Map<String, String>> {
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Map<String, String> map(ResultSet r, int columnNumber, StatementContext ctx) throws SQLException {
+        return (Map<String, String>) r.getObject(columnNumber);
```
Comment (qualidafial) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreColumnMapper.java]: It's unfortunate we can't use `getObject(columnNumber, Map.class)` here for some added type safety.

**Comment by qualidafial** [comment]
We use `hstore` at my work, so I can't wait to use this.

File: postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java (hunk 1 above)
Comment (arteam) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: Wow, I didn't see that! Neat that it's already implemented in a much better way!

File: postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * An argument factory which binds Java's {@link Map} to Postgres' hstore type.
+ */
+public class HStoreArgumentFactory implements ArgumentFactory {
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Optional<Argument> build(Type type, Object value, StatementContext ctx) {
+        if (value instanceof Map) {
```
Comment (arteam) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: Unfortunately, we don't have a generic information about the type. The type which is passed to argument factories is derived from the value by a simple `getClass` call, which returns only the erased type.

File: postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java (hunk 2 above)
Comment (arteam) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: Thanks! I solved the issue by delegating this task to the Postgres JDBC driver, which already implements converting with escaping.

File: postgres/src/main/java/org/jdbi/v3/postgres/HStoreColumnMapper.java (hunk 3 above)
Comment (arteam) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreColumnMapper.java]: Yeah, I think the Postgres driver could actually support `hstore` for this method as well.

File: postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java (hunk 4 above)
Comment (qualidafial) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: It depends.

SQL objects should pass the full type information from the method parameter (or bean property, in the case of `@BindBean`).

However if you're using the fluent interface, then `type` could be `Map.class`, `Object.class` (for nulls), or `Map<String,String>` depending on which method you used to bind the parameter.

Comment (arteam) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: Hmm, I didn't consider the issue from this angle. Let me check and get back to you.

Comment (arteam) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: I gave a little bit thought to this question, and I think we should not limit API to the `Map<String, String>` type. The PostgreSQL JDBC driver supports any maps during inserts, and we are just proxy calls to it. By removing the restriction on the `<String, String>` generic parameter we support all types of insertions in JDBI and provide a good developer experience.

Comment (qualidafial) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: Fair enough.

Can we still change this `instanceof` condition to `Map.class.isAssignableFrom(GenericTypes.getErasedType(type)`? That will allow this argument factory to also capture nulls passed to Map arguments in SQL objects.

Comment (arteam) [inline postgres/src/main/java/org/jdbi/v3/postgres/HStoreArgumentFactory.java]: Sure, I've updated the PR.

**Pull request merged**